pyro/infer/mcmc/hmc.py

```diff
@@ -0,0 +1,225 @@
+    return z_next, r_next
+
+
+def retain_grads(z):
+    for key in z:
+        # XXX: can be removed with PyTorch 0.3
+        if z[key].is_leaf and not z[key].requires_grad:
+            z[key].requires_grad = True
+        z[key].retain_grad()
+
+
+def detach_nodes(z):
+    for key in z:
+        z[key] = Variable(z[key].data, requires_grad=True)
+
+
+class HMC(TracePosterior):
+    def __init__(self,
+                 model,
+                 step_size=0.5,
+                 num_steps=3,
+                 warmup_steps=20,
+                 num_samples=1000):
+        self.model = model
+        self.step_size = step_size
+        self.num_steps = num_steps
+        self.num_samples = num_samples
+        self.warmup_steps = warmup_steps
+        # simulation run attributes - will be set in self._setup
+        # at start of run
+        self._cached_param_grads = {}
+        self._args = None
+        self._kwargs = None
+        self._accept_cnt = None
+        self._prototype_trace = None
+        super(HMC, self).__init__()
+
+    def _log_prob(self, z):
+        """
+        Return log pdf of the model with sample sites replayed from z_trace
+        """
+        z_trace = self._prototype_trace.copy()
+        for name, value in z.items():
+            z_trace.nodes[name]['value'] = value
+        model_trace = poutine.trace(poutine.replay(self.model, trace=z_trace)) \
+            .get_trace(*self._args, **self._kwargs)
+        return model_trace.log_pdf()
+
+    def _grad_potential(self, z):
+        log_joint_prob = self._log_prob(z)
+        log_joint_prob.backward()
+        grad_potential = {}
+        for name, value in z.items():
+            grad_potential[name] = -value.grad.clone().detach()
+            grad_potential[name].volatile = False
+        return grad_potential
+
+    def _energy(self, z, r):
+        kinetic_energy = 0.5 * torch.sum(torch.stack([r[name] ** 2 for name in r]))
+        potential_energy = - self._log_prob(z)
+        return kinetic_energy + potential_energy
+
+    def _setup(self, *args, **kwargs):
+        self._accept_cnt = 0
+        self._args = args
+        self._kwargs = kwargs
+        # set the trace prototype to inter-convert between trace object
+        # and dict object used by the integrator
+        self._prototype_trace = poutine.trace(self.model).get_trace(*args, **kwargs)
+        for name, node in self._prototype_trace.iter_stochastic_nodes():
+            if not node['fn'].reparameterized:
+                raise ValueError('Found non-reparameterized node in the model at site: {}'.format(name))
+        prototype_trace_log_pdf = self._prototype_trace.log_pdf().data[0]
+        if np.isnan(prototype_trace_log_pdf) or np.isinf(prototype_trace_log_pdf):
+            raise ValueError('Model specification incorrect - trace log pdf is NaN, Inf or 0.')
+        # store the current value of param gradients so that they
```

Comment:
See above; this hacking would be simplified if we changed `poutine.replay()` to replay `param` sites.